dev/DynamicDependency/MddLifetimeManagement.cpp

```diff
@@ -0,0 +1,185 @@
+                if (!g_test_ddlmPackageNamePrefix.empty())
+                {
+                    const auto packageId{ appExtension.Package().Id() };
+                    std::wstring name{ packageId.Name().c_str() };
+                    if ((name.rfind(g_test_ddlmPackageNamePrefix.c_str(), 0) != 0) ||
+                        (CompareStringOrdinal(packageId.PublisherId().c_str(), -1, g_test_ddlmPackagePublisherId.c_str(), -1, TRUE) != CSTR_EQUAL))
+                    {
+                        // The package's Name prefix or PublisherId don't match the expected value. Skip it
+                        continue;
+                    }
+                }
+
+                // appExtension.Id == "ddlm-<major.minor.build.revision>-<architecture>"
+                const auto id{ appExtension.Id() };
+                PACKAGE_VERSION version{};
+                WCHAR architectureAsString[9 + 1]{};
+                const auto maxIdLength{ ARRAYSIZE(L"ddlm-12345.12345.12345.12345.abcdefghi") - 1 }; // -1 for length not counting null-terminator
+                if ((id.size() >= maxIdLength) ||
+                    (swscanf_s(id.c_str(), L"ddlm-%hu.%hu.%hu.%hu-%9s", &version.Major, &version.Minor, &version.Build, &version.Revision, architectureAsString, static_cast<unsigned>(ARRAYSIZE(architectureAsString))) != 5))
+                {
+                    (void)LOG_HR_MSG(ERROR_INVALID_DATA, "%ls", id.c_str());
+                    continue;
+                }
+
+                // Found one
+                auto packageFullName{ appExtension.Package().Id().FullName().c_str() };
+                ddlmPackages.push_back(MddCore::LifetimeManagement::DDLMPackage(packageFullName, version));
+            }
+
+            // Did we find more than one?
+            if (ddlmPackages.size() <= 1)
+            {
+                // Nothing to remove
+                continue;
+            }
+
+            // Sort the list by version in descending order to simplify using it
+            std::sort(ddlmPackages.begin(), ddlmPackages.end(), std::greater<MddCore::LifetimeManagement::DDLMPackage>());
+
+            // What's the highest version with a healthy status
+            auto keeper{ ddlmPackages.begin() };
+            for (; keeper != ddlmPackages.end(); ++keeper)
+            {
+                auto package{ packageManager.FindPackageForUser(currentUser, keeper->PackageFullName()) };
+                auto status{ package.Status() };
+                if (status.VerifyIsOK())
+                {
+                    break;
+                }
+            }
+            const auto keeperIndex{ keeper - ddlmPackages.begin() };
+
+            // Remove all older packages (best effort)
+            for (++keeper; keeper != ddlmPackages.end(); ++keeper)
+            {
+                const auto n{ keeper - ddlmPackages.begin() };
+                wprintf(L"Boo %I64u %I64u\n", keeperIndex, n);
```

Comment:
Maybe log instead of wprintf?

Reply:
Leftover debugging crud (the 'Boo' is the giveaway :P)